fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/AgilityContext.java

```diff
@@ -0,0 +1,238 @@
+
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.StampedLock;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Create floating sub contexts from a caller context and commit when they reach time/write quota.
+ */
+public interface AgilityContext {
+
+    static AgilityContext factory(FDBRecordContext callerContext, boolean useAgileContext) {
+        return useAgileContext ? new Agile(callerContext) : new NonAgile(callerContext);
+    }
+
+    // `apply` should be called when a returned value is expected
+    <R> CompletableFuture<R> apply(Function<FDBRecordContext, CompletableFuture<R>> function) ;
+
+    // `accept` should be called when a returned value is not expected
+    void accept(Consumer<FDBRecordContext> function);
+
+    // `set` should be called for writes - keeping track of write size (if needed)
+    void set(byte[] key, byte[] value);
+
+    void flush();
+
+    default CompletableFuture<byte[]> get(byte[] key) {
+        return apply(context -> context.ensureActive().get(key));
+    }
+
+    default void clear(byte[] key) {
+        accept(context -> context.ensureActive().clear(key));
+    }
+
+    default void clear(Range range) {
+        accept(context -> context.ensureActive().clear(range));
+    }
+
+
+    /**
+     * A floating window (agile) context - create sub contexts and commit them as they reach their time/size quota.
+     */
+    class Agile implements AgilityContext {
+
+        private final FDBRecordContextConfig.Builder contextConfigBuilder;
+        private final FDBDatabase database;
+        private final FDBRecordContext callerContext; // for counters updates only
+
+        private FDBRecordContext currentContext;
+        private long creationTime;
+        private int currentWriteSize;
+        private final long timeQuotaMillis;
+        private final long sizeQuotaBytes;
+        // Lock plan:
+        //   apply/accept - use read lock, release it within the future
+        //   create context - synced and under the read lock of apply/accept
+        //   commitNow - use write lock to ensure exclusivity
+        // also:
+        //   create and commit context functions are each synchronized
+        //   committingNow boolean is used to prevent threads cluttering at the commit function when a quota is reached
+        private final StampedLock lock = new StampedLock();
+        private final Object createLockSync = new Object();
+        private final Object commitLockSync = new Object();
+        private boolean committingNow = false;
+
+        Agile(FDBRecordContext callerContext) {
+            this.callerContext = callerContext;
+            contextConfigBuilder = callerContext.getConfig().toBuilder();
+            contextConfigBuilder.setWeakReadSemantics(null); // Since this context may be used for retries, do not allow weak read semantic
```

Comment:
We don't actually do retries in here, right? We still want to set this to `null`, but that's because we don't want all the transactions to use the same read-version, or they'll definitely conflict with themselves.

Reply:
Changed comment. 
✅